File: LeetCode/Go/15-two-pointers-三数之和.go
/*
 * @lc app=leetcode.cn id=15 lang=golang
 *
 * [15] 三数之和
 */
package leetcode

// @lc code=start
import "sort"

func threeSum(nums []int) [][]int {
	var res = [][]int{}
	sort.Ints(nums)
	for i := 0; i < len(nums)-2; i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}
		var l, r = i + 1, len(nums) - 1
		for l < r {
			var sum = nums[i] + nums[l] + nums[r]
			if sum == 0 {
				res = append(res, []int{nums[i], nums[l], nums[r]})
				for l+1 < r && nums[l] == nums[l+1] {
					l++
				}
				for l < r-1 && nums[r] == nums[r-1] {
					r--
				}
				l, r = l+1, r-1
			} else if sum < 0 {
				l++
			} else {
				r--
			}
		}
	}
	return res
}

// @lc code=end

File: LeetCode/Java/15-two-pointers-三数之和.java
/*
 * @lc app=leetcode.cn id=15 lang=java
 *
 * [15] 三数之和
 */

// @lc code=start
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        int n = nums.length;

        Arrays.sort(nums);

        for (int t = 0; t < n - 2; t++) {
            if (nums[t] > 0)
                return res;
            if (t > 0 && nums[t] == nums[t - 1])
                continue;

            int l = t + 1;
            int r = n - 1;

            while (l < r) {
                int sum = nums[t] + nums[l] + nums[r];
                if (sum == 0) {
                    res.add(Arrays.asList(nums[t], nums[l], nums[r]));
                    while (l + 1 < r && nums[l] == nums[l + 1])
                        ++l;
                    while (l < r - 1 && nums[r] == nums[r - 1])
                        --r;
                    ++l;
                    --r;
                } else if (sum < 0)
                    ++l;
                else
                    --r;
            }
        }

        return res;
    }
}
// @lc code=end
